### PR TITLE
Abort when duplicate sample names are found #143

### PR DESCRIPTION
**Pull request by nitrozyna**
**Abort when duplicate sample names are found #143 solution comments**

I had a few issues compiling the project, in particular two test cases failing. Since I thought they weren't key to my implementation of the task, I commented them out.

Failed tests:
  CreateDatabaseIndexesStepTest.testIndexesAreCreated
  StatisticsMongoWriterTest.shouldCreateIndexesInCollection

I implemented the check for repeated sample names with a HashSet data structure. The reason why I used a Hashset is because HashSet is much faster than TreeSet (constant-time versus log-time for most operations) but offers no ordering guarantees which as far as I understand does not seem to be a requirement in the pipeline.

I couldn't find any suitable exceptions under src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/ hence I decided to throw an IOException since that best describes the encountered problem.

I created a new folder to test on under the following name: same_sample_names.vcf.gz which contains a duplicate of sample: HG00096.

I also created a new test case in VcfHeaderReaderTest.java to test for duplicate samples.

Improvement thoughts:
If required, I could find all the duplicate samples, since for now the IOException is thrown on the first duplicate sample encountered.

### CODE REVIEW

File: src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java [hunk 1]
```diff
@@ -98,19 +98,33 @@ public VariantSourceEntity read() throws Exception {
...
+            }
+            uniformSamples.add(sample);
+        }
         return new VariantSourceEntity(source);
+
     }
 
     @Override
     public void open(ExecutionContext executionContext) throws ItemStreamException {
+
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: Remove blank lines at beginning and end of methods.

File: src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
 import org.springframework.core.io.Resource;
 
 import uk.ac.ebi.eva.commons.models.data.VariantSourceEntity;
-
+import java.util.*;
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: Use explicit imports following our internal coding guidelines. Can be configured in the IDE.

File: src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java [hunk 3]
```diff
@@ -98,19 +98,33 @@ public VariantSourceEntity read() throws Exception {
...
+    private VariantSourceEntity doRead() throws IOException {
         if (variantReader == null) {
             throw new IllegalStateException("The method VcfHeaderReader.open() should be called before reading");
         }
         variantReader.pre();
         source.addMetadata(VARIANT_FILE_HEADER_KEY, variantReader.getHeader());
+
+        List<String> sampleNames = variantReader.getSampleNames();
+
+        Set<String> uniformSamples = new HashSet<String>();
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: A shorter way would be to create the Set using the List as input, and comparing the sizes. To retrieve (possibly multiple) duplicate samples, you can use other methods in those classes.

I would also suggest renaming this variable to `uniqueSamples`, because that expresses the idea a bit better than `uniform`.

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java [hunk 4]
```diff
@@ -48,6 +51,20 @@
     @Rule
     public PipelineTemporaryFolderRule temporaryFolderRule = new PipelineTemporaryFolderRule();
 
+    @Test(expected = IOException.class)
+    public void testDuplicates() throws Exception {
```
Comment (cyenyxe) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: `testDuplicateSamples` to make clear that you are not referring to, for instance, duplicate records.

File: src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java [hunk 5]
```diff
@@ -98,19 +98,33 @@ public VariantSourceEntity read() throws Exception {
...
+
+        for (String sample : sampleNames) {
+            if ((uniformSamples.contains(sample))) {
+                throw new IOException("Duplicate sample name given: " + sample);
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: As you mention in the PR comments there was no suitable exception yet to cover this. But this one is too generic, because it could be thrown due to any issue with access to disk, etc. Maybe you could create a `DuplicateSamplesFoundException` inside this package (it wouldn't make sense to add it to `runner`)?

**Comment by cyenyxe** [comment]
It seems the test that expects and exception is now failing because it is not thrown.

**Comment by nitrozyna** [comment]
Sorry, I haven't finished yet. Once I do, I will submit a pull request again.

File: src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/DuplicateSamplesFoundException.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2015-2017 EMBL - European Bioinformatics Institute
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/DuplicateSamplesFoundException.java]: Copyright 2018, because that is the year the file was created in.

File: src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java [hunk 7]
```diff
@@ -100,12 +104,22 @@ public VariantSourceEntity read() throws Exception {
...
         if (variantReader == null) {
             throw new IllegalStateException("The method VcfHeaderReader.open() should be called before reading");
         }
         variantReader.pre();
         source.addMetadata(VARIANT_FILE_HEADER_KEY, variantReader.getHeader());
+
+        List<String> sampleNames = new ArrayList<String> ( variantReader.getSampleNames());
+        HashSet<String> uniqueSampleNames = new HashSet<String>(sampleNames);
+
+        if (sampleNames.size() != uniqueSampleNames.size()){
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: Missing white space before the `{`

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java [hunk 8]
```diff
@@ -48,6 +51,19 @@
...
 
+    @Test(expected = DuplicateSamplesFoundException.class)
+    public void testDuplicateSamples() throws Exception {
+        File input = getResource(INPUT_FILE_PATH_DUPLICATES);
+
+        VariantStudy.StudyType studyType = VariantStudy.StudyType.COLLECTION;
+        VariantSource.Aggregation aggregation = VariantSource.Aggregation.NONE;
+
+        VcfHeaderReader headerReader = new VcfHeaderReader(input, FILE_ID, STUDY_ID, STUDY_NAME,
+                studyType, aggregation);
```
Comment (cyenyxe) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: I noticed the method below does the same as this one, but others below use the constants. I think the code would look more clear if we use directly the constants; could you please also change the method right below?

Comment (nitrozyna) [reply src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: Could you elaborate? I am not sure if I understand what you mean here. I thought I am using constants the same way as other methods.

Comment (nitrozyna) [reply src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: Upon further revision I think what you requested is that I remove the indirection through the "studyType" and "aggregation" locally scoped variables for VariantStudy.StudyType.COLLECTION; and  VariantSource.Aggregation.NONE; respectively.

I initially took the indirect route as the fully qualified constant names are rather verbose and appear multiple times ( in the setup and final assertions ). Additionally these benefit from enhanced visibility of being set once in the test preamble. 

One compromise would be to adopt the same convention as that used for FILE_ID , STUDY_ID, etc. 

That is :  
Declare a set of private constants in the Test class and used these as indirections to the fully qualified constants. This has several benefits:
            - static final fields won't take up extra stack space per test 
            - the compiler will likely eliminate the intermediate storage anyway 
            - the developers can quickly asses and change the studyType and aggregation if required

Could you please confirm if you would prefer the private static field approach , or just the one that is established in the final test?

Comment (cyenyxe) [reply src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: Oops, you were right in using the indirections. I didn't notice the asserts below were using them.

Comment (nitrozyna) [reply src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: That's alright, thanks! I changed it back.

File: src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/DuplicateSamplesFoundException.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.ac.ebi.eva.pipeline.runner.exceptions;
+import java.util.List;
+
+/**
+ * Exception thrown where a string value argument is not recognized or valid in a function.
```
Comment (jmmut) [inline src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/DuplicateSamplesFoundException.java]: Is this message correct? looks unrelated to me

File: src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java [hunk 10]
```diff
@@ -100,12 +104,22 @@ public VariantSourceEntity read() throws Exception {
...
         }
         variantReader.pre();
         source.addMetadata(VARIANT_FILE_HEADER_KEY, variantReader.getHeader());
+
+        List<String> sampleNames = new ArrayList<String> ( variantReader.getSampleNames());
+        HashSet<String> uniqueSampleNames = new HashSet<String>(sampleNames);
+
+        if (sampleNames.size() != uniqueSampleNames.size()) {
+            for (String str : uniqueSampleNames) {
+              sampleNames.remove(str);
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: While this is functional, the performance complexity of remove is O(n). Since this remove operation is inside a loop, complexity here is O(n*n). This can escalate quickly even for studies with moderately large number of samples (say 1000 genomes). @cyenyxe Any thoughts?

Comment (nitrozyna) [reply src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java]: Fair point! Since a HashSet is usually backed by a HashMap I could indeed just check for duplication while inserting into the HashSet rather than by removing from the slower ArrayList.
HashSet's add(obj) returns a status boolean showing whether the object was a duplicate or not. So I can just iteratively add the sample names rather than use a copy constructor and then check for any true flags returned at the end. Does that sound reasonable?

File: src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/DuplicateSamplesFoundException.java (hunk 9 above)
Comment (nitrozyna) [reply src/main/java/uk/ac/ebi/eva/pipeline/runner/exceptions/DuplicateSamplesFoundException.java]: Well spotted, thanks!
